chainer/functions/loss/triplet.py

```diff
@@ -46,29 +47,37 @@ def forward(self, inputs):
             loss = xp.sum(self.dist_hinge) / N
         else:
             loss = self.dist_hinge
+
+        self.retain_inputs((0, 1, 2))
         return xp.array(loss, dtype=numpy.float32),
 
-    def backward(self, inputs, gy):
-        xp = cuda.get_array_module(*inputs)
+    def backward(self, indexes, grad_outputs):
+        anchor, positive, negative = self.get_retained_inputs()
 
-        anchor, positive, negative = inputs
         N = anchor.shape[0]
-
         x_dim = anchor.shape[1]
+
+        xp = cuda.get_array_module(anchor)
         tmp = xp.repeat(self.dist_hinge[:, None], x_dim, axis=1)
```

Comment:
`backward` method must be implemented in `chainer.Variable` level with `chainer.functions` instead of NumPy/CuPy level operations so that `Triplet` supports double backward, so we can not use `self.dist_hinge` here but need to compute it again with the retained inputs (or implement as a grad FunctionNode, `TripletGrad`, so that we can implement the gradient in NumPy/CuPy level).

Reply:
We only use `dist_hinge` to mask the upstream gradients, i.e. I don't thinks it is necessary to wrap them in `Variables` since there are no gradients being propagated. Please correct me if I am wrong.

Reply:
Okay. I got what you intend and it is well considered. It should work, but I still have a bit concern if this style is valid in  `backward` method. Can I confirm it to @beam2d ?

Reply:
I confirmed that to @beam2d and it is okay as you say that the masking itself makes no effect on propagation. Sorry for bothering you.